src/v/cluster/partition_leaders_table.cc

```diff
@@ -149,63 +175,81 @@ void partition_leaders_table::update_partition_leader(
             }
             // reset the term for new ntp revision
             if (
               revision_id_valid
-              && revision_id > it->second.partition_revision) {
-                it->second.update_term = model::term_id{};
+              && revision_id > p_it->second.partition_revision) {
+                p_it->second.update_term = model::term_id{};
             }
         }
 
-        // existing partition
-        if (it->second.update_term > term) {
+        if (p_it->second.update_term > term) {
             vlog(
               clusterlog.trace,
               "skip update for partition {} with previous term {} current term "
               "{}",
               ntp,
               term,
-              it->second.update_term);
+              p_it->second.update_term);
             // Do nothing if update term is older
             return;
         }
+        /**
+         * Update leader less partition counter when leadership changed
+         *
+         * no leader -> leader - decrement leaderless counter
+         * leader -> no leader - increment leaderless counter
+         */
+        if (p_it->second.current_leader && !leader_id) {
+            ++_leaderless_partition_count;
+        } else if (!p_it->second.current_leader && leader_id) {
+            --_leaderless_partition_count;
+        }
 
         // if current leader has value, store it as a previous leader
-        if (it->second.current_leader) {
-            it->second.previous_leader = it->second.current_leader;
+        if (p_it->second.current_leader) {
+            p_it->second.previous_leader = p_it->second.current_leader;
         }
-        it->second.current_leader = leader_id;
-        it->second.update_term = term;
+        p_it->second.current_leader = leader_id;
+        p_it->second.update_term = term;
         if (revision_id_valid) {
-            it->second.partition_revision = revision_id;
+            p_it->second.partition_revision = revision_id;
+        }
+    } else {
+        if (!leader_id) {
+            ++_leaderless_partition_count;
         }
+        /**
+         * We only increment version if any of the maps content was modified
+         */
+        ++_version;
     }
+
     vlog(
       clusterlog.trace,
       "updated partition: {} leader: {{term: {}, current leader: {}, previous "
       "leader: {}, revision: {}}}",
       ntp,
-      it->second.update_term,
-      it->second.current_leader,
-      it->second.previous_leader,
-      it->second.partition_revision);
+      p_it->second.update_term,
+      p_it->second.current_leader,
+      p_it->second.previous_leader,
+      p_it->second.partition_revision);
     // notify waiters if update is setting the leader
     if (!leader_id) {
         return;
     }
     // update stable leader term
-    it->second.last_stable_leader_term = term;
+    p_it->second.last_stable_leader_term = term;
 
     if (auto it = _leader_promises.find(ntp); it != _leader_promises.end()) {
```

Comment:
From my understanding of the code I don't see where we would early out so if nothing has changed then we still reach this line? This somehow seems surprising though, why would we need to notify anyone about this then?

Do we also need to make `_leader_promises` hierarchical?